config/default.yml

```diff
@@ -929,6 +929,13 @@ RSpec/SpecFilePathSuffix:
     - "**/spec/**/*"
   Reference: https://www.rubydoc.info/gems/rubocop-rspec/RuboCop/Cop/RSpec/SpecFilePathSuffix
 
+RSpec/StringAsInstanceDoubleConstant:
+  Description: Do not use a string as `instance_double` constant.
+  Enabled: pending
+  Safe: false
```

Comment:
What are those unsafe cases?

Reply:
In real-world-rspec, I could find:
```
                                      ^^^^^^^^
/Users/pirj/source/real-world-rspec/engine/spec/views/locomotive/current_site/edit_spec.rb:44:39: C: [Correctable] RSpec/StringAsInstanceDoubleConstant: Do not use a string as instance_double constant.
      let(:_policy) { instance_double('policy', create?: true, point?: true, update_advanced?: true, destroy?: false) }
```

and
```
/Users/pirj/source/real-world-rspec/chef/spec/unit/provider/package/openbsd_spec.rb:56:31: C: [Correctable] RSpec/StringAsInstanceDoubleConstant: Do not use a string as instance_double constant.
              instance_double("shellout", stdout: "#{name}-#{version}\n")
                              ^^^^^^^^^^
```

But this is probably due to verified doubles being turned off.
I'd say this should become a regular `double`, or become a proper class constant like in:
```
/Users/pirj/source/real-world-rspec/chef/chef-config/spec/unit/config_spec.rb:930:33: C: [Correctable] RSpec/StringAsInstanceDoubleConstant: Do not use a string as instance_double constant.
          cmd = instance_double("Mixlib::ShellOut", exitstatus: 0, stdout: locales, error!: nil)
                                ^^^^^^^^^^^^^^^^^^
```

Reply:
Yes, I think we can't mark it as safe for autocorrect, because to do so would require too many assumptions to be met about the test setup. 

The other concern mentioned in the original repo issue here is that using the class to initialize requires actually loading the class (and again I'm not making any assumptions about Zeitwerk etc for the end user's test suite here). 

Although we can't be sure the class is loaded to safely autocorrect, much of the benefit of this Cop is enforcing that the `instance_double` really does work as intended (because this way the class is loaded in order to initialize)

From the docs:
>Constructs a test double against a specific class. If the given class name has been loaded, only instance methods defined on the class are allowed to be stubbed. In all other ways it behaves like a [double](https://rubydoc.info/github/rspec/rspec-mocks/double).
Source: https://rubydoc.info/github/rspec/rspec-mocks/RSpec%2FMocks%2FExampleMethods:instance_double

Reply:
Do you want me to add a sentence like "This is not marked as safe for autocorrect, since the correction only works when the class is loaded" ?

Reply:
I'm hoping that I was on point with this, and I added it. Happy to amend as needed!

Reply:
What are the unsafe cases?

Reply:
I do get one.

> using the class to initialize requires actually loading the class

This makes sense. We certainly don’t want to enforce loading classes.

From the doc:
> If the given class name has been loaded, only instance methods defined on the class are allowed to be stubbed. In all other ways it behaves like a [double](https://rubydoc.info/github/rspec/rspec-mocks/double).

So the two legit usages from my perspective are:
 - instance_double with a class
 - double with a string when the intention is to NOT load a class

I’m struggling to quickly come up with an idea how to provide clear separation between those two for our ysers.

Reply:
There’s a dedicated @satefy tag, it will be parsed and added to docs. Please keep the description short.

I apologise for brief response- I’m from my phone.

Reply:
Perfect. Added the @safety tag! I think this is a lot more clearly presented now, with just enough info folks won't need to hunt down the original issue or even the RSpec docs.

Reply:
Yes, I think the key thing is that this Cop will enforce that an `instance_double` really is a validating double. String instantiation "falls back" to a simple non-verifying double when the class has not been loaded. So the behavior for strings is ambiguous in that it may result in a validating double or a non-validating double. And if there are other setup issues, that could actually be a dynamic problem, and lead to flaky tests.

Reply:
Nice summary!

This makes me think is if the cop itself is safe, it's just the autocorrection to a literal class that is not. The `SafeAutocorrect: false` setting should do the job of highlighting the issue, and letting users decide whether they want to switch to a regular double and keep a string to prevent loading, or to benefit from what instance_double offers, and switch to the class. @bquorning WDYT?